src/commands/deploy.ts

```diff
@@ -0,0 +1,279 @@
+      Setting up database ...
+      ✔ DONE
+      
+      Copying development ...
+      ✔ DONE
+      
+      Restarting app ...
+      ✔ DONE
+      
+      Deploy completed. Good bye!
+`;
+
+export const args = {
+  "--app": {
+    type: AppArg,
+    alias: "-a",
+  },
+  "--force": Boolean,
+} satisfies ArgsSpec;
+
+export enum Action {
+  DEPLOY_ANYWAYS = "Deploy anyways",
+  SYNC_ONCE = "Sync once",
+  CANCEL = "Cancel (Ctrl+C)",
+}
+
+const AppDeploymentStepsToAppDeployState = (step: string | undefined): string => {
+  switch (step) {
+    case "NOT_STARTED":
+      return "Deploy not started";
+    case "STARTING":
+    case "BUILDING_ASSETS":
+    case "UPLOADING_ASSETS":
+      return "Building frontend assets";
+    case "CONVERGING_STORAGE":
+      return "Setting up database";
+    case "PUBLISHING_TREE":
+      return "Copying development";
+    case "RELOADING_SANDBOX":
+      return "Restarting app";
+    case "COMPLETED":
+      return "Deploy completed";
+    default:
+      return "Unknown step";
+  }
+};
+
+enum AppDeploymentSteps {
+  NOT_STARTED = "NOT_STARTED",
+  STARTING = "STARTING",
+  BUILDING_ASSETS = "BUILDING_ASSETS",
+  UPLOADING_ASSETS = "UPLOADING_ASSETS",
+  CONVERGING_STORAGE = "CONVERGING_STORAGE",
+  PUBLISHING_TREE = "PUBLISHING_TREE",
+  RELOADING_SANDBOX = "RELOADING_SANDBOX",
+  COMPLETED = "COMPLETED",
+}
+
+/**
+ * Runs the deploy process.
+ */
+
+export const command = (async (ctx, firstRun = true) => {
+  const spinner = ora();
+  let prevProgress: string | undefined = AppDeploymentStepsToAppDeployState("NOT_STARTED");
+  let action: Action;
+
+  const filesync = await FileSync.init({
+    user: await getUserOrLogin(),
+    dir: ctx.args._[0],
+    app: ctx.args["--app"],
+  });
+
+  const log = filesync.log.extend("deploy");
+
+  if (firstRun) {
+    log.println(
+      boxen(
+        sprint`
+        ggt v${config.version}
+  
+        App         ${filesync.app.slug}
+        Editor      https://${filesync.app.slug}.gadget.app/edit
+        Playground  https://${filesync.app.slug}.gadget.app/api/graphql/playground
+        Docs        https://docs.gadget.dev/api/${filesync.app.slug}
+    
+        Endpoints ${
+          filesync.app.hasSplitEnvironments
+            ? `
+          • https://${filesync.app.primaryDomain}
+          • https://${filesync.app.slug}--development.gadget.app`
+            : `
+          • https://${filesync.app.primaryDomain}`
+        }
+        `,
+        {
+          padding: 1,
+          borderStyle: "round",
+          dimBorder: true,
+        },
+      ),
+    );
+  }
+
+  const { localHashes, gadgetHashes } = await filesync._getHashes();
+
+  const upToDate = isEqualHashes(localHashes, gadgetHashes);
+
+  if (!upToDate) {
+    log.println(`
+    ${""}
+    ${chalk.bold(
+      `Local files have diverged from remote. Run a sync once to converge your files or keep ${chalk.italic(
+        "ggt sync",
+      )} running in the background.`,
+    )}
+    `);
```

Comment:
There are 2 ways to use the `log.println` method:

1. ```ts
   log.println("some string")
    ```
2. ```ts
   log.println`some string`
    ```

The first way is useful when you want to print a simple message without any formatting or a string that's already been formatted. The second way is when you want to print a message with formatting since it uses [`chalk-template`](https://www.npmjs.com/package/chalk-template) under the hood.

Also, there are multiple `log.print` methods and one of them prints new lines before and after the message called `printlns`.

---

With all of that in mind, we can simplify this code to this and it will look the same:

```ts
log.printlns`
	Local files have diverged from remote. Run a sync once to converge your files or keep {italic ggt sync} running in the background.
`;
```

What do you think?

Reply:
😮 ya that looks cleaner!! good idea!